src/lib/support/jsontlv/TlvToJson.cpp

```diff
@@ -0,0 +1,356 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/support/CHIPMemString.h"
+#include "lib/support/ScopedBuffer.h"
+#include <json/json.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/Base64.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/jsontlv/ElementTypes.h>
+#include <lib/support/jsontlv/TlvToJson.h>
+
+namespace chip {
+
+namespace {
+
+constexpr uint32_t kTemporaryImplicitProfileId = 0x1122;
+
+/// RAII to switch the implicit profie id for a reader
+class ImplicitProfileIdChange
+{
+public:
+    ImplicitProfileIdChange(TLV::TLVReader & reader, uint32_t id) : mReader(reader), mOldImplicitProfileId(reader.ImplicitProfileId)
+    {
+        reader.ImplicitProfileId = id;
+    }
+    ~ImplicitProfileIdChange() { mReader.ImplicitProfileId = mOldImplicitProfileId; }
+
+private:
+    TLV::TLVReader & mReader;
+    uint32_t mOldImplicitProfileId;
+};
+
+const char * GetJsonElementStrFromType(const ElementTypeContext & ctx)
+{
+    switch (ctx.tlvType)
+    {
+    case TLV::kTLVType_UnsignedInteger:
+        return kElementTypeUInt;
+    case TLV::kTLVType_SignedInteger:
+        return kElementTypeInt;
+    case TLV::kTLVType_Boolean:
+        return kElementTypeBool;
+    case TLV::kTLVType_FloatingPointNumber:
+        return ctx.isDouble ? kElementTypeDouble : kElementTypeFloat;
+    case TLV::kTLVType_ByteString:
+        return kElementTypeBytes;
+    case TLV::kTLVType_UTF8String:
+        return kElementTypeString;
+    case TLV::kTLVType_Null:
+        return kElementTypeNull;
+    case TLV::kTLVType_Structure:
+        return kElementTypeStruct;
+    case TLV::kTLVType_Array:
+        return kElementTypeArray;
+    default:
+        return kElementTypeEmpty;
+    }
+};
+
+/*
+ * Encapsulates the element information required to constract Json element name string in a Json object.
+ *
+ * The generated Json element name string is constracted as:
+ *     'TagNumber:ElementType-SubElementType'.
+ */
+struct JsonObjectElementContext
+{
+    JsonObjectElementContext(TLV::TLVReader & reader)
+    {
+        tag          = reader.GetTag();
+        type.tlvType = reader.GetType();
+        if (type.tlvType == TLV::kTLVType_FloatingPointNumber)
+        {
+            type.isDouble = reader.IsElementDouble();
+        }
+    }
+
+    std::string GenerateJsonElementName() const
+    {
+        std::string str = std::to_string(TLV::TagNumFromTag(tag));
```

Comment:
So this setup completely fails at the claimed use case of "encode attribute IDs as profile tags".....  I know there's a followup issue tracking that, but please do note that it's completely broken so far.

Reply:
Updated logic to detect implicit profile IDs and treat them separately from the existing encoding of generic profiles (so that TagNumFromTag is consistent).

This should make the code at least generate things that are distinguishable or error out if internal consistency is bad.

Reply:
Updated encoding a bit to not lose profile and not try to get tag number for explicitly not supported cases. it will not distinguish between "context tag 1" and "implicit profile tag 1" however that seems format-intended.

The paths that are new (profile/tag and the ???) should never be exercised, but at least we have them for easy debuggin of odd code,